### PR TITLE
Properly translate C++ exception to Python exception when creating Python buffer from wrapped object

### PR DESCRIPTION
**Pull request by vnlitvinov**
## Description

* Refactor exception translation a little - extract inline translation to a callable helper;
* Call the helper if an exception occurs when trying to get a buffer from wrapped object;
* Added the test, which crashes with `Fatal Python error` without the fix (unhandlable from within Python) and passes after it is done.

Closes: #2764, #3336

## Suggested changelog entry:

```rst
* Properly translate C++ exception to Python exception when creating Python buffer from wrapped object
```

### CODE REVIEW

**Comment by rwgk** [comment]
I need to find a block of time to look carefully.

But high-level: There is already too much stuff in internals.h, adding in the exception translator code seems very unfortunate.

pybind11 is poorly factored in general, I'd love to not make it worse, even though it can get tricky sometimes.

Could you please try to move the exception translator code into its own header, maybe pybind11/detail/exception_translators.h?

You'll have to update a couple files when adding a new header. An easy way to pin-point the locations is e.g.:

```
$ git grep detail/value_and_holder.h
CMakeLists.txt:    include/pybind11/detail/value_and_holder.h
tests/extra_python_package/test_files.py:    "include/pybind11/detail/value_and_holder.h",
```

File: include/pybind11/detail/class.h
Comment (vnlitvinov) [reply include/pybind11/detail/class.h:588]: indeed, applied

File: tests/test_buffers.cpp
Comment (vnlitvinov) [reply tests/test_buffers.cpp:173]: agreed

File: tests/test_buffers.cpp
Comment (vnlitvinov) [reply tests/test_buffers.cpp:177]: removed

File: include/pybind11/detail/class.h
Comment (rwgk) [reply include/pybind11/detail/class.h:590]: Just from inspecting [the sources](https://github.com/pybind/pybind11/blob/66c3774a6402224b1724329c81c880e76633a92b/include/pybind11/pybind11.h#L1740), I'm thinking `info` could be `nullptr` here, which would lead to segfaults further down.

Does that make sense to you?

If true, it is an existing problem, but I'd lean towards fixing it in this PR while we're at it. Simply (untested):

```
if (info == nullptr) {
    pybind11_fail("FATAL UNEXPECTED SITUATION: tinfo->get_buffer() returned nullptr.");
}
```

File: include/pybind11/detail/exception_translation.h
Comment (rwgk) [inline include/pybind11/detail/exception_translation.h:4]: Just `2024`?

Also maybe an extra `s` at the end of the line above? (... translate ... excpetions to ... exceptions)

**Review by rwgk — APPROVED** [review]
Looks good to me, thanks!

The suggestions are really minor.

File: include/pybind11/detail/class.h
Comment (vnlitvinov) [reply include/pybind11/detail/class.h:590]: this would also crash the app immediately... but I guess it's better to at least be explicit about the reason, so added

File: include/pybind11/detail/exception_translation.h
Comment (vnlitvinov) [reply include/pybind11/detail/exception_translation.h]: fixed

File: include/pybind11/detail/class.h
Comment (rwgk) [reply include/pybind11/detail/class.h:590]: Thanks, yes, much better. I'm used to looking a traceback hundreds of levels deep. Seeing or not seeing that exception can make the difference between an immediate aha and several days lost.

**Review by rwgk — APPROVED** [review]
@henryiii @EthanSteinberg @Skylion007: Do you have ~10 minutes for a second review?

**Review by EthanSteinberg — APPROVED** [review]
This looks good to me, nice job!

(As a side note though, get_buffer seems to have a way more convoluted design than it should ...)

**Comment by rwgk** [comment]
Thanks a lot @EthanSteinberg for the review!

**Pull request merged**
